node/network/collator-protocol/src/lib.rs

```diff
@@ -71,28 +71,25 @@ impl From<util::validator_discovery::Error> for Error {
 
 type Result<T> = std::result::Result<T, Error>;
 
-enum ProtocolSide {
+/// What side of the collator protocol is being engaged
+pub enum ProtocolSide {
+	/// Validators operate on the relay chain.
 	Validator(validator_side::Metrics),
+	/// Collators operate on a parachain.
 	Collator(CollatorId, collator_side::Metrics),
 }
 
 /// The collator protocol subsystem.
 pub struct CollatorProtocolSubsystem {
-	protocol_side: ProtocolSide, 
+	protocol_side: ProtocolSide,
 }
 
 impl CollatorProtocolSubsystem {
 	/// Start the collator protocol.
 	/// If `id` is `Some` this is a collator side of the protocol.
-	/// If `id` is `None` this is a validator side of the protocol. 
+	/// If `id` is `None` this is a validator side of the protocol.
 	/// Caller must provide a registry for prometheus metrics.
-	pub fn new(id: Option<CollatorId>, registry: Option<&prometheus::Registry>) -> Self {
-		use metrics::Metrics;
-		let protocol_side = match id {
-			Some(id) => ProtocolSide::Collator(id, collator_side::Metrics::register(registry)),
-			None => ProtocolSide::Validator(validator_side::Metrics::register(registry)),
-		};
-
+	pub fn new(protocol_side: ProtocolSide) -> Self {
```

Comment:
There was a discussion about this refactoring https://github.com/paritytech/polkadot/pull/1741#discussion_r492054076.
I guess it's fine for now, but needs to be changed in future for cumulus integration.

Reply:
Back in that discussion, nobody came up with a substantially better version. I don't mind changing it in the future if there's a better idea, but as an inevitable consequence of having two distinct sides to the implementation, we do have to construct one or the other of them.